File: prisoners.js
// run: node prisoners.js < prisoners.txt

function solve(left, right) {
	var key = left + "-" + right;
	if (cache[key] != null) return cache[key]
	if (right <= left) return cache[key] = 0;

	var min = null;
	for (var i = 0; i < list.length; i++) {
		if (list[i] >= left && list[i] <= right) {
			var gold = {
				value: right - left,
				left:  solve(left, list[i] - 1),
				right: solve(list[i] + 1, right)
			}
			var sum = gold.value + gold.left + gold.right;
			min = Math.min(min || sum, sum);
		}
	}
	min = min || 0;
	return cache[key] = min;
}

function parse(data) {
	var lines = data.split("\n");
	var T = parseInt(lines[0]);
	for (var t = 1; t <= T; t++) {
		var tokens = lines[2 * t - 1].split(" ");
		var prisoner = parseInt(tokens[0]);
		list = lines[2 * t].split(" ");
		for (var i = 0; i < list.length; i++)
			list[i] = parseInt(list[i]);
		cache = [];
		var result = solve(1, prisoner);
		console.log("Case #" + t + ": " + result);
	}
}

var cache = [];
var list = [];
var buffer = "";

process.stdin.setEncoding('utf8');
process.stdin.on("readable", function() {
	var chunk = process.stdin.read();
	buffer += chunk == null ? "" : chunk;
});
process.stdin.on("end", function() {
	parse(buffer);
});

File: prisoners.c
// compile: gcc prisoners.c
// run:     ./a.out < prisoners.txt

#include <stdio.h>
#define MAX 10000

int cache[MAX+1][MAX+1];
int a[MAX];
int N, P;

int solve(int L, int R) {
	if (cache[L][R] == -1) {
		int min = 0;
		for (int i = 0; i < N; i++) {
			if (a[i] >= L && a[i] <= R) {
				int cost = R-L;
				int costL = solve(L, a[i]-1);
				int costR = solve(a[i]+1, R);
				int sum = cost + costL + costR;
				if (min == 0)
					min = sum;
				if (min > sum)
					min = sum;
			}
		}
		cache[L][R] = min;
		return min;
	} else {
		return cache[L][R];
	}
}

int main() {
	int T;
	scanf("%d\n", &T);
	for (int t = 1; t <= T; t++) {
		int result = 0;
		scanf("%d %d\n", &P, &N);
		for (int i = 0; i < N; i++) {
			scanf("%d", &a[i]);
		}
		for (int i = 0; i <= P; i++) {
			for (int j = 0; j <= P; j++) {
				cache[i][j] = -1;
			}
		}
		result = solve(1, P);
		printf("Case #%d: %d\n", t, result);
	}
	return 0;
}
